### PR TITLE
Optimize `OffsetList`

### PR DESCRIPTION
**Pull request by frankmcsherry**
Use an optimized `OffsetList` in `DatumBatch`. Remove key-specialized arrangements that *should* now not result in space savings due to the optimized representation.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/compute/src/row_spine.rs [hunk 1]
```diff
@@ -208,15 +207,15 @@ mod container {
...
+            &self.storage[*lower..*upper]
         }
         fn len(&self) -> usize {
             self.offsets.len() - 1
         }
 
         fn with_capacities(item_cap: usize, byte_cap: usize) -> Self {
             // TODO: be wary of `byte_cap` greater than 2^32.
-            let mut offsets = OffsetList::with_capacity(item_cap + 1);
+            let mut offsets = crate::row_spine::OffsetOptimized::with_capacity(item_cap + 1);
```
Comment (frankmcsherry) [inline src/compute/src/row_spine.rs:218]: The `OffsetOptimized::with_capacity` call ignores its argument. Perhaps we should think about that to be sure we like that behavior.

File: src/compute/src/render/context.rs [hunk 2]
```diff
@@ -1072,55 +996,14 @@ where
...
-                return (SpecializedArrangement::RowUnit(oks), errs);
-            }
-        } else {
-            // We expect to always have type information for arrangements, do getting here is
-            // likely a bug.
-            tracing::error!(%name, "missing types for arrangement");
-        }
-
+    ) -> (MzArrangement<S>, Collection<S, DataflowError, i64>) {
         // Catch-all: Just use RowRow.
```
Comment (antiguru) [inline src/compute/src/render/context.rs:1000]: Stale comment.

**Review by antiguru — APPROVED** [review]
Looks good. There are some remaining stale comments, but I think we can merge now and fix comments later.

**Comment by frankmcsherry** [comment]
Sounds good. Merging now so we get a bit of time with it in `main` before the next release cut. Thanks very much for the review!

**Pull request merged**